uast/original.go

```diff
@@ -0,0 +1,241 @@
+		return nil, ErrUnsupported.New("top level object as root node")
+	}
+
+	if len(src) > 1 {
+		return nil, ErrUnexpectedObjectSize.New(1, len(src))
+	}
+
+	for _, obj := range src {
+		return c.toNode(obj)
+	}
+
+	panic("not reachable")
+}
+
+func (c *BaseToNoder) toNode(obj interface{}) (*Node, error) {
+	m, ok := obj.(map[string]interface{})
+	if !ok {
+		return nil, ErrUnexpectedObject.New("map[string]interface{}", obj)
+	}
+
+	n := NewNode()
+	for k, o := range m {
+
+		switch ov := o.(type) {
+		case map[string]interface{}:
+			child, err := c.mapToNode(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, child)
+		case []interface{}:
+			children, err := c.sliceToNodes(k, ov)
+			if err != nil {
+				return nil, err
+			}
+
+			n.Children = append(n.Children, children...)
+		default:
+			if err := c.addProperty(n, k, o); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sort.Sort(byOffset(n.Children))
+	return n, nil
+}
+
+func (c *BaseToNoder) mapToNode(k string, obj map[string]interface{}) (*Node, error) {
+	n, err := c.toNode(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	n.Properties[InternalRoleKey] = k
+	return n, nil
+}
+
+func (c *BaseToNoder) sliceToNodes(k string, s []interface{}) ([]*Node, error) {
+	var ns []*Node
+	for _, v := range s {
+		n, err := c.toNode(v)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Properties[InternalRoleKey] = k
+		ns = append(ns, n)
+	}
+
+	return ns, nil
+}
+
+func (c *BaseToNoder) addProperty(n *Node, k string, o interface{}) error {
+	switch {
+	case c.isTokenKey(k):
+		if n.Token != "" {
+			return fmt.Errorf("two token keys for same node: %s", k)
+		}
+
+		n.Token = fmt.Sprint(o)
+	case c.InternalTypeKey == k:
+		s := fmt.Sprint(o)
+		if err := c.setInternalKey(n, s); err != nil {
+			return err
+		}
+
+		tk := c.syntheticToken(s)
+		if tk != "" {
+			if n.Token != "" {
+				return fmt.Errorf("two token keys for same node: %s", k)
+			}
+
+			n.Token = tk
+		}
+	case c.OffsetKey == k:
+		i, err := toUint32(o)
+		if err != nil {
+			return err
+		}
+
+		n.StartPosition.Offset = i
+	case c.LineKey == k:
+		i, err := toUint32(o)
+		if err != nil {
+			return err
+		}
+
+		n.StartPosition.Line = i
+	default:
+		n.Properties[k] = fmt.Sprint(0)
+	}
+
+	return nil
+}
+
+func (c *BaseToNoder) isTokenKey(key string) bool {
+	return c.TokenKeys != nil && c.TokenKeys[key]
+}
+
+func (c *BaseToNoder) syntheticToken(key string) string {
+	if c.SyntheticTokens == nil {
+		return ""
+	}
+
+	return c.SyntheticTokens[key]
+}
+
+func (c *BaseToNoder) setInternalKey(n *Node, k string) error {
+	if n.InternalType != "" {
+		return fmt.Errorf("two internal keys for same node: %s, %s",
+			n.InternalType, k)
+	}
+
+	n.InternalType = k
+	return nil
+}
+
+func toUint32(v interface{}) (uint32, error) {
+	switch o := v.(type) {
+	case string:
+		i, err := strconv.ParseUint(o, 10, 32)
+		if err != nil {
+			return 0, err
+		}
+
+		return uint32(i), nil
+	case uint32:
```

Comment:
if this comes from a JSON-parsed `map[string]interface{}` the only possibility here is `int64` because that's the only integer type JSON handles when it does not know the type.

`map[string]interface{}`, `[]interface{}`, `string`, `bool`, `int64` and `float64` are the only types that can come out of a JSON unmarshalled to `map[string]interface{}`.